### PR TITLE
Feature 4

### PR DESCRIPTION
**Pull request by denis**
The user can drag shapes from the palette and drop them onto the drawing canvas.

Build: http://barushev.net/shaper/builds/4/

What's done:
- Learned about React state.
- Learned about React refs to components.
- Learned about HTML Drag and Drop API.
- Learned about TypeScript default-initialized parameters.

### CODE REVIEW

File: src/components/Canvas.tsx [hunk 1]
```diff
@@ -1,12 +1,70 @@
...
+          this.state.shapes.map((shape) => {
+            return this.renderShape(shape);
+          })
+        }
+      </svg>
     </div>;
   }
+
+  private renderShape(shape: models.Shape) {
+    let key = Math.random();
```
Comment (denis) [inline src/components/Canvas.tsx:35]: I'm not sure that this is a good idea.

Comment (danschultz) [inline src/components/Canvas.tsx:35]: It's better use a key value that is unique to the shape. The key is used to determine if this is the same component or not when rerendering.

File: src/components/Canvas.tsx [hunk 2]
```diff
@@ -1,12 +1,70 @@
...
+
+    if (shapeType == "circle") {
+      shape = new models.Circle(centerX, centerY, size);
+    } else if (shapeType == "square") {
+      shape = new models.Square(centerX, centerY, size);
+    } else if (shapeType == "triangle") {
+      shape = new models.Triangle(centerX, centerY, size);
+    }
+
+    this.setState({shapes: this.state.shapes.concat(shape)});
```
Comment (danschultz) [inline src/components/Canvas.tsx:68]: Tip: You can use destructuring as a shortcut for concat.

``` ts
this.setState({shapes: [...this.state.shapes, shape]});
```

File: src/components/Canvas.tsx (hunk 1 above)
Comment (danschultz) [inline src/components/Canvas.tsx:35]: Here's the docs on keys and performance: https://facebook.github.io/react/docs/reconciliation.html#keys

**Comment by danschultz** [comment]
LGTM

**Pull request merged**